### PR TITLE
perf Adding performance co pilot tests

### PR DESCRIPTION
**Pull request by nasastry**
Adding Performance Co Pilot(PCP) test cases

Signed-off-by: Nageswara R Sastry <rnsastry@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by nasastry** [comment]
Test results on Power9:

```
JOB ID     : 6e1b151150727ab903dc9e13bad11edd50ca4ef1
JOB LOG    : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-03-27T11.52-6e1b151/job.log
 (1/5) perf_pcp.py:PCP.test_pmcd_daemon: PASS (0.78 s)
 (2/5) perf_pcp.py:PCP.test_pmda_perfevent_install: PASS (5.72 s)
 (3/5) perf_pcp.py:PCP.test_pcp_cmd: PASS (0.81 s)
 (4/5) perf_pcp.py:PCP.test_perfevent_24x7_events: PASS (4.80 s)
 (5/5) perf_pcp.py:PCP.test_pmda_perfevent_remove: PASS (4.14 s)
RESULTS    : PASS 5 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 17.17 s
```

Test results on Power8:

```
JOB ID     : d09ec8921e4fb59e8a2f684c6f4fc46a5f9bc2b6
JOB LOG    : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-03-27T11.53-d09ec89/job.log
 (1/5) perf_pcp.py:PCP.test_pmcd_daemon: PASS (0.89 s)
 (2/5) perf_pcp.py:PCP.test_pmda_perfevent_install: PASS (7.17 s)
 (3/5) perf_pcp.py:PCP.test_pcp_cmd: PASS (0.94 s)
 (4/5) perf_pcp.py:PCP.test_perfevent_24x7_events: CANCEL: Not supported on Power8 (0.82 s)
 (5/5) perf_pcp.py:PCP.test_pmda_perfevent_remove: PASS (4.45 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 1
JOB TIME   : 15.16 s
```

**Comment by nasastry** [comment]
Added `guest check` and `LPAR` check.

**Comment by nasastry** [comment]
After the new update: 

Power9 
```
JOB ID     : a8f12eacf37fd8243421c54b58eed6d99a958438
JOB LOG    : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-03-27T13.22-a8f12ea/job.log
 (1/5) perf_pcp.py:PCP.test_pmcd_daemon: PASS (0.77 s)
 (2/5) perf_pcp.py:PCP.test_pmda_perfevent_install: PASS (5.71 s)
 (3/5) perf_pcp.py:PCP.test_pcp_cmd: PASS (0.81 s)
 (4/5) perf_pcp.py:PCP.test_perfevent_24x7_events: PASS (5.08 s)
 (5/5) perf_pcp.py:PCP.test_pmda_perfevent_remove: PASS (4.15 s)
RESULTS    : PASS 5 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 17.35 s
```

Power8:
```
JOB ID     : f4194b2ccf7e3cab728d6e132e2c7dcd6f91776d
JOB LOG    : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-03-27T13.22-f4194b2/job.log
 (1/5) perf_pcp.py:PCP.test_pmcd_daemon: PASS (1.04 s)
 (2/5) perf_pcp.py:PCP.test_pmda_perfevent_install: PASS (6.42 s)
 (3/5) perf_pcp.py:PCP.test_pcp_cmd: PASS (1.12 s)
 (4/5) perf_pcp.py:PCP.test_perfevent_24x7_events: CANCEL: Not supported on Power8 (1.07 s)
 (5/5) perf_pcp.py:PCP.test_pmda_perfevent_remove: PASS (4.56 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 1
JOB TIME   : 15.18 s
```

File: perf/perf_pcp.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    def test_pmcd_daemon(self):
+        output = process.run("systemctl start pmcd", shell=True)
+        output = process.run("systemctl status pmcd", shell=True)
+        if "active (running)" not in output.stdout:
+            self.fail("PCP: Can not start pmcd daemon")
+
+    def test_pmda_perfevent_install(self):
+        if not os.path.isfile("/var/lib/pcp/pmdas/perfevent/Install"):
+            self.fail("PCP: perfevent pmda package not installed")
+        cmd = "cd /var/lib/pcp/pmdas/perfevent/; echo pipe | ./Install"
```
Comment (harish-24) [inline perf/perf_pcp.py]: Better to use `os.chdir()` read file with python and pass to `./Install` script.

File: perf/perf_pcp.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            self.cancel("Not supported on Power8")
+        # -s 2 collects two samples
+        cmd = "pmval -s 2 perfevent.hwcounters.hv_24x7.PM_MBA0_CLK_CYC.value"
+        output = process.run(cmd, shell=True)
+        # -s 2 collects two samples
+        cmd = "pmval -s 2 perfevent.hwcounters.hv_24x7.PM_PB_CYC.value"
+        output = process.run(cmd, shell=True)
+
+    def test_pmda_perfevent_remove(self):
+        cmd = "cd /var/lib/pcp/pmdas/perfevent/; echo pipe | ./Remove"
```
Comment (harish-24) [inline perf/perf_pcp.py]: same here.

File: perf/perf_pcp.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        if "Cannot connect to PMCD" in output.stdout:
+            self.fail("PCP: pmcd daemon dead")
+
+    def test_perfevent_24x7_events(self):
+        self.lpar_24x7_ppc64le_check()
+        if self.cpu_arch == 'power8':
+            self.cancel("Not supported on Power8")
+        # -s 2 collects two samples
+        cmd = "pmval -s 2 perfevent.hwcounters.hv_24x7.PM_MBA0_CLK_CYC.value"
+        output = process.run(cmd, shell=True)
```
Comment (harish-24) [inline perf/perf_pcp.py]: `output` variable not required?

File: perf/perf_pcp.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        # -s 2 collects two samples
+        cmd = "pmval -s 2 perfevent.hwcounters.hv_24x7.PM_PB_CYC.value"
+        output = process.run(cmd, shell=True)
```
Comment (harish-24) [inline perf/perf_pcp.py]: here too anything to be checked with `output`?

File: perf/perf_pcp.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+        else:
+            self.cancel("PCP is not supported on %s" % dist.name)
+
+        for pkg in pkgs:
+            if not smg.check_installed(pkg) and not smg.install(pkg):
+                self.cancel("Package %s is missing/could not be installed"
+                            % pkg)
+
+    def lpar_24x7_ppc64le_check(self):
+        if 'ppc' not in process.system_output("uname -p", ignore_status=True):
```
[the remark this answers is not in the document]
Comment (nasastry) [reply perf/perf_pcp.py]: @harish-24 Ok

File: perf/perf_pcp.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            self.cancel("Processor is not ppc64")
+        # Check if this is a guest
+        # 24x7 is not suported on guest
+        cpu_output = process.run("cat /proc/cpuinfo")
```
[the remark this answers is not in the document]
Comment (nasastry) [reply perf/perf_pcp.py]: @harish-24 `process.system_output` also runs `process.run`. With `process.system_output` some times I am seeing issue with `utf-8` characters.

File: perf/perf_pcp.py (hunk 1 above)
Comment (nasastry) [reply perf/perf_pcp.py]: @harish-24 when `os.chdir()` used need to add `try ... catch` code. `process.run` code is better than `adding os.chdir and process.run`

File: perf/perf_pcp.py (hunk 3 above)
Comment (nasastry) [reply perf/perf_pcp.py]: @harish-24 `output` content is not required.

File: perf/perf_pcp.py (hunk 4 above)
Comment (nasastry) [reply perf/perf_pcp.py]: @harish-24 `output` content is not required.

File: perf/perf_pcp.py (hunk 2 above)
Comment (nasastry) [reply perf/perf_pcp.py]: @harish-24 when os.chdir() used need to add try ... catch code. process.run code is better than adding os.chdir and process.run

File: perf/perf_pcp.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (harish-24) [reply perf/perf_pcp.py]: @nasastry Anyhow as `process.system_output` uses `process.run` then `utf-8` issue should be seen everywhere. If `cpu_output.stdout` is all that is needed system_output is explicitly provided for the same purpose.

File: perf/perf_pcp.py (hunk 1 above)
Comment (harish-24) [reply perf/perf_pcp.py]: @nasastry As this is a avocado/python based framework it is better to have maximum possible code in python to reduce execution time and moreover builtin functions are already available. Also adding `try...catch` does not add any overhead.

**Comment by nasastry** [comment]
@harish-24 Changed code according to your comments. Please review.

Power9 results:
```
JOB ID     : 45197115079e10cd3defbe49d2381a4519c8d1f6
JOB LOG    : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-03-29T10.24-4519711/job.log
 (1/5) perf_pcp.py:PCP.test_pmcd_daemon: PASS (0.76 s)
 (2/5) perf_pcp.py:PCP.test_pmda_perfevent_install: PASS (6.90 s)
 (3/5) perf_pcp.py:PCP.test_pcp_cmd: PASS (0.86 s)
 (4/5) perf_pcp.py:PCP.test_perfevent_24x7_events: PASS (4.81 s)
 (5/5) perf_pcp.py:PCP.test_pmda_perfevent_remove: PASS (4.16 s)
RESULTS    : PASS 5 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 18.30 s
```

Power8 results:
```
JOB LOG    : /home/workspace/runAvocadoFVTTest/avocado-fvt-wrapper/results/job-2019-03-29T10.25-78d8cc0/job.log
 (1/5) perf_pcp.py:PCP.test_pmcd_daemon: PASS (1.55 s)
 (2/5) perf_pcp.py:PCP.test_pmda_perfevent_install: PASS (7.25 s)
 (3/5) perf_pcp.py:PCP.test_pcp_cmd: PASS (0.96 s)
 (4/5) perf_pcp.py:PCP.test_perfevent_24x7_events: CANCEL: Not supported on Power8 (0.82 s)
 (5/5) perf_pcp.py:PCP.test_pmda_perfevent_remove: PASS (4.45 s)
RESULTS    : PASS 4 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 1
JOB TIME   : 15.84 s
```

File: perf/perf_pcp.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        # -s 2 collects two samples
+        cmd = "pmval -s 2 perfevent.hwcounters.hv_24x7.PM_MBA0_CLK_CYC.value"
+        process.run(cmd, shell=True)
+        # -s 2 collects two samples
+        cmd = "pmval -s 2 perfevent.hwcounters.hv_24x7.PM_PB_CYC.value"
+        process.run(cmd, shell=True)
+
+    def test_pmda_perfevent_remove(self):
+        os.chdir("/var/lib/pcp/pmdas/perfevent/")
+        cmd = "./Remove"
```
Comment (harish-24) [inline perf/perf_pcp.py]: Is `echo pipe` removed cautiously here?

Comment (nasastry) [reply perf/perf_pcp.py]: @harish-24 For remove `pipe` is not required.

**Comment by PraveenPenguin** [comment]
@harish-24 thanks for review

**Pull request merged**